tidy3d/components/mode/simulation.py

```diff
@@ -0,0 +1,337 @@
+"""Defines mode solver simulation class."""
+
+from __future__ import annotations
+
+from typing import List, Tuple
+
+import numpy as np
+import pydantic.v1 as pd
+
+from ...exceptions import SetupError, ValidationError
+from ...log import log
+from ..base import cached_property
+from ..boundary import BoundarySpec, PECBoundary
+from ..grid.grid import Grid
+from ..grid.grid_spec import GridSpec
+from ..mode_spec import ModeSpec
+from ..monitor import ModeMonitor, ModeSolverMonitor
+from ..simulation import AbstractYeeGridSimulation, Simulation
+from ..source import GaussianPulse, ModeSource, SourceTime
+from ..structure import Structure
+from ..types import Axis, Direction, FreqArray
+from ..validators import validate_freqs_min, validate_freqs_not_empty
+
+# dummy run time for conversion to FDTD sim
+# should be very small -- otherwise, generating tmesh will fail or take a long time
+RUN_TIME = 1e-30
+
+# Maximum allowed size of the field data produced by the mode solver
+MAX_MODES_DATA_SIZE_GB = 20
+
+MODE_MONITOR_NAME = "<<<MODE_SOLVER_MONITOR>>>"
+
+
+class ModeSimulation(AbstractYeeGridSimulation):
+    """Mode solver simulation.
+
+    Example
+    -------
+    >>> from tidy3d import C_0, ModeSpec
+    >>> lambda0 = 1
+    >>> freq0 = C_0 / lambda0
+    >>> freqs = [freq0]
+    >>> sim_size = lambda0, lambda0, 0
+    >>> mode_spec = ModeSpec(num_modes=4)
+    >>> sim = ModeSimulation(
+    ...     size=sim_size,
+    ...     freqs=freqs,
+    ...     mode_spec=mode_spec
+    ... )
+
+    """
+
+    mode_spec: ModeSpec = pd.Field(
+        ...,
+        title="Mode specification",
+        description="Container with specifications about the modes to be solved for.",
+    )
+
+    freqs: FreqArray = pd.Field(
+        ..., title="Frequencies", description="A list of frequencies at which to solve."
+    )
+
+    direction: Direction = pd.Field(
+        "+",
+        title="Propagation direction",
+        description="Direction of waveguide mode propagation along the axis defined by its normal "
+        "dimension.",
+    )
+
+    colocate: bool = pd.Field(
+        True,
+        title="Colocate fields",
+        description="Toggle whether fields should be colocated to grid cell boundaries (i.e. "
+        "primal grid nodes). Default is ``True``.",
+    )
+
+    boundary_spec: BoundarySpec = pd.Field(
+        BoundarySpec.all_sides(PECBoundary()),
+        title="Boundaries",
+        description="Specification of boundary conditions along each dimension. If ``None``, "
+        "PML boundary conditions are applied on all sides. NOTE: for Mode simulations, this "
+        "is required to be PECBoundary on all sides. The 'ModeSpec' can be used to apply PML.",
+    )
+
+    monitors: Tuple[None, ...] = pd.Field(
```

Comment:
should we consider instead removing Monitors and sources from the abstract case and then adding them back into the Simulation and EMESimulation? (same with `.data`)

Reply:
yeah, I think that makes sense

Reply:
they have to be individually defined for each simulation anyway, since each type of simulation accepts different types of sources and monitors

Reply:
well, but they all do have a lot in common, in terms of the geometric properties and basic api for monitors, so maybe we should leave it for the time being

Reply:
I see, I'll leave it up to you, but we could consider making the base classes use `Literal[()]` for these fields perhaps? and then if the subclasses add sources / monitors, we can define them as `Tuple[SpecificTypes, ...]`? just a thought.

Reply:
This still doesn't seem to work:

```
E           pydantic.v1.error_wrappers.ValidationError: 2 validation errors for ModeSimulation
E           sources
E             unexpected value; permitted:  (type=value_error.const; given=(); permitted=())
E           monitors
E             unexpected value; permitted:  (type=value_error.const; given=(); permitted=())
```